// trs/trscore.js
var os = require('os');
var sleep = require('sleep');
function Sleep(seconds) {
	sleep.sleep(seconds);
}
function Messages(suite, test){
	this.suite=suite;
	this.test=test;
}
Messages.prototype.ShowTest = function() {
	return "Suite #"+this.suite+":"+"Test #"+this.test+" : ";
};
Messages.prototype.ResultOk = function(process) {
	process.send({ msg: this.ShowTest()+"Result OK" });
};
Messages.prototype.Fail = function(process, msg) {
	process.send({ msg: this.ShowTest()+"Fail: "+msg });
};
Messages.prototype.Error = function(process, msg) {
	process.send({ msg: this.ShowTest()+"Error: "+msg });
};
exports.Messages = Messages;

function Manager() {
	if (os.platform()=="win32") {
		this.file = process.argv[1].slice(-0, -7) +"suite.xml";
		var manageapp = require(process.argv[2]+"/lib/Win/ApplicationManagementWin");
		var xml2js = require(process.argv[2]+'/xml2js/node_modules/xml2js');
		this.manage = new manageapp.ApplicationManagement();
		var fileSystem = require("fs");
		var parser = new xml2js.Parser(); 
        var extendedData;
        parser.parseString(fileSystem.readFileSync(this.file).toString(),function (err, result) {
        	extendedData=result;
        });
        this.pause=Number(extendedData.suite.test[0].repeat[0].$.pause);
        this.repeat=Number(extendedData.suite.test[0].repeat[0]._);
        this.manage.SetAppName(String(extendedData.suite.application));
        this.manage.SetWindowName(String(extendedData.suite.windowName));
        this.msg = new Messages(extendedData.suite.$.name, extendedData.suite.test[0].$.name);
    }
	else {
		console.log("Operating system not supported");
		process.exit();
	}
}	
Manager.prototype.Test = function (fun){
	fun(this);
}
Manager.prototype.MouseClickAt = function (button, x, y){
	this.manage.SetMousePos(x,y);
	this.manage.MouseClick(button);
	Sleep(1);
}
Manager.prototype.Shortcut = function (command, key){
	Sleep(1);
	this.manage.KeyDown(Number(command));
	this.manage.KeyPress(Number(key));
	this.manage.KeyUp(Number(command));
	Sleep(1);
}

global.Manager = Manager;
global.Sleep = Sleep;
