// trs/trscore.js
var os = require('os');
var sleep = require('sleep');
function Sleep(miliseconds) {
	sleep.usleep(miliseconds*1000);
}
function Messages(suite, test){
	this.suite=suite;
	this.test=test;
	this.sleepTime=500;
}
Messages.prototype.ResultOk = function(process) {
	process.send({suite: this.suite, test: this.test, msg:" OK" });
};
Messages.prototype.Fail = function(process, msg) {
	process.send({suite: this.suite, test: this.test, msg:" Fail: "+msg });
};
Messages.prototype.Error = function(process, msg) {
	process.send({ msg:" Error: "+msg });
};
Messages.prototype.Msg = function(process, msg) {
	process.send({suite: this.suite, test: this.test, msg:this.test+" Msg: "+msg });
};
exports.Messages = Messages;

function Manager() {
	if (os.platform()=="win32") {
		var manageapp = require(process.argv[1].slice(-0, -9)+"/lib/Win/ApplicationManagementWin");
		this.manage = new manageapp.ApplicationManagement();
        this.pause=0;
        this.repeat=0;
        this.msg = new Messages("suite","test");
    }
	else {
		console.log("Operating system not supported");
		process.exit();
	}
}
Manager.prototype.Test = function (fun){
	fun(this);
}
Manager.prototype.ResultOk = function (process, msg){
	this.msg.ResultOk(process);
}
Manager.prototype.Fail = function (process, msg){
	this.msg.Fail(process, msg);
}
Manager.prototype.Error = function (process, msg){
	this.msg.Error(process, msg);
}
Manager.prototype.Msg = function (process, msg){
	this.msg.Msg(process, msg);
}
Manager.prototype.SetWindowSize = function (w, h){
	this.kofX=this.GetScreenWidth()/w;
	this.kofY=this.GetScreenHeight()/h;	
}
Manager.prototype.MouseClickAt = function (button, x, y){
	this.manage.SetMousePos(this.kofX*x,this.kofY*y);
	this.manage.MouseClick(button);
	Sleep(500);
}
Manager.prototype.Shortcut = function (command, key){
	Sleep(500);
	this.manage.KeyDown(Number(command));
	this.manage.KeyPress(Number(key));
	this.manage.KeyUp(Number(command));
	Sleep(500);
}
Manager.prototype.SetAppName = function (name){
	this.manage.SetAppName(String(name));
}
Manager.prototype.SetWindowName = function (name){
	this.manage.SetWindowName(String(name));
}
Manager.prototype.StartApp = function (){
	this.manage.StartApp();
	Sleep(500);
	this.manage.SetActive();
	Sleep(500);
}
Manager.prototype.CloseApp = function (){
	this.manage.CloseApp();
	Sleep(500);
}
Manager.prototype.WindowMinimize = function (){
	this.manage.WindowMinimize();
	Sleep(500);
}
Manager.prototype.WindowMaximize = function (){
	this.manage.WindowMaximize();
	Sleep(500);
}
Manager.prototype.SetActive = function (){
	this.manage.SetActive();
	Sleep(500);
}
Manager.prototype.WindowRestore = function (){
	this.manage.WindowRestore();
	Sleep(500);
}
Manager.prototype.GetScreenWidth = function (){
	return this.manage.GetScreenWidth();
}
Manager.prototype.GetScreenHeight = function (){
	return this.manage.GetScreenHeight();
}
Manager.prototype.KeyDown = function (key){
	this.manage.KeyDown(Number(key));
	Sleep(500);
}
Manager.prototype.KeyUp = function (key){
	this.manage.KeyUp(Number(key));
	Sleep(500);
}
Manager.prototype.KeyPress = function (command, key){
	this.manage.KeyPress(Number(key));
	Sleep(500);
}
Manager.prototype.SetMousePos = function (x, y){
	this.manage.SetMousePos(this.kofX*x,this.kofY*y);
	Sleep(500);
}
Manager.prototype.MouseMove = function (x, y, pause){
	this.manage.SetMousePos(this.kofX*x,this.kofY*y);
	Sleep(500);
	this.manage.MouseMove(this.kofX*x,this.kofY*y, pause);
	Sleep(500);
}
Manager.prototype.MouseDown = function (button){
	this.manage.MouseDown(button);
	Sleep(500);
}
Manager.prototype.MouseUp = function (button){
	this.manage.MouseUp(button);
	Sleep(500);
}
Manager.prototype.MouseClick = function (button){
	this.manage.MouseClick(button);
	Sleep(500);
}
Manager.prototype.MouseWheelDown = function (){
	this.manage.MouseWheelDown();
	Sleep(500);
}
Manager.prototype.MouseWheelUp = function (){
	this.manage.MouseWheelUp();
	Sleep(500);
}
Manager.prototype.MouseWheelLeft = function (){
	this.manage.MouseWheelLeft();
	Sleep(500);
}
Manager.prototype.MouseWheelRight = function (){
	this.manage.MouseWheelRight();
	Sleep(500);
}
Manager.prototype.PrintScreen = function (path, name){
	this.manage.PrintScreen(path+name);
	Sleep(500);
}
Manager.prototype.PrintText = function (text){
	text = text.toUpperCase();
	this.manage.KeyDown(Number(16));
	for(var i=0; i<text.length;i++){
		var key = String(text).charCodeAt(i);
		if(key >= 48 && key <= 90) {
			this.manage.KeyPress(String(text).charCodeAt(i));
		}
	}
	this.manage.KeyPress(49);
	this.manage.KeyUp(Number(16));
	Sleep(500);
}

global.Manager = Manager;
global.Sleep = Sleep;
